src/capwap_loc_provider.erl

```diff
@@ -0,0 +1,151 @@
+%% Copyright (C) 2013-2023, Travelping GmbH <info@travelping.com>
+
+%% This program is free software: you can redistribute it and/or modify
+%% it under the terms of the GNU Affero General Public License as published by
+%% the Free Software Foundation, either version 3 of the License, or
+%% (at your option) any later version.
+
+%% This program is distributed in the hope that it will be useful,
+%% but WITHOUT ANY WARRANTY; without even the implied warranty of
+%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+%% GNU Affero General Public License for more details.
+
+%% You should have received a copy of the GNU Affero General Public License
+%% along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+-module(capwap_loc_provider).
+
+-behavior(gen_server).
+
+%% API
+-export([start/1, start_link/1]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2]).
+
+%% user api
+-export([flush_loc_cache/0, get_loc/1, get_loc/2, load_config/1]).
+
+-include_lib("kernel/include/logger.hrl").
+
+-define(CACHE_LOC, capwap_loc_provider_cache).
+-define(CACHE_CONFIG, capwap_loc_config_cache).
+-define(TIMER_WAIT, 2000).
+
+-record(loc_cache, {loc_tab, config_tab, refresh, providers, provider_chain = undefined}).
+
+%% callbacks
+
+-callback config_fun(any()) -> fun((<<>>) -> {error, any()} | {location, <<>>, <<>>}).
+%% A separate callback returning an opaque ref would be
+%% useful in the case of stateful providers.
+%% This option is not covered yet.
+
+%% user api
+
+start_link(Config) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, Config, []).
+
+start(Config) ->
+    gen_server:start({local, ?MODULE}, ?MODULE, Config, []).
+
+flush_loc_cache() ->
+    gen_server:call(?MODULE, flush_loc_cache).
+
+load_config(Config) ->
+    gen_server:call(?MODULE, {load_config, Config}).
+
+%% Uses exceptions and error (for the provider itself)
+get_loc(Name) ->
+    get_loc(Name, true).
+
+get_loc(Name, true) ->
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    case ets:lookup(?CACHE_LOC, Name) of
+	[{Name, Loc = {location, _, _}}] -> Loc;
+	[] -> gen_server:call(?MODULE, {get_loc, true, Name}, 20000)
+    end;
+get_loc(Name, false) ->
+    ?LOG(debug, "bypassing cache"),
+    gen_server:call(?MODULE, {get_loc, false, Name}, 20000).
+
+%% gen_server impl
```

Comment:
This is impl of validation - not of gen server - please or move this comment or remove it at all